lxd/device/pci/pci.go

```diff
@@ -158,3 +158,32 @@ func deviceProbeWait(pciDev Device) error {
 
 	return fmt.Errorf("Device took too long to activate at %q", driverPath)
 }
+
+// NormaliseAddress converts common PCI address notation to the kernel's notation.
+func NormaliseAddress(addr string) string {
+	// Ignore empty values.
+	if addr == "" {
+		return addr
+	}
+
+	// PCI devices can be specified as "0000:XX:XX.X" or "XX:XX.X".
+	// However, the devices in /sys/bus/pci/devices use the long format which
+	// is why we need to make sure the prefix is present.
+	if len(addr) == 7 {
+		addr = fmt.Sprintf("0000:%s", addr)
+	}
+
+	// Ensure all addresses are lowercase.
+	addr = strings.ToLower(addr)
+
+	return addr
+}
+
+// DeviceExists checks whether a configured PCI device exists.
+func DeviceExists(addr string) error {
+	if addr != "" && !shared.PathExists(fmt.Sprintf("/sys/bus/pci/devices/%s", addr)) {
```

Comment:
@stgraber same with this one, the empty check is device specific and seems like it should be in each device (depending on whether it is optional or not).

Reply:
For that one it's making `DeviceExists` a one-line function and adding 4 lines of boiler plate to all callers so that doesn't seem worth it.

Instead I'll revert to having a `validatePCIDevice` in device_utils_generic.go as we did before and have that take the address directly.

Reply:
I suppose either way it will return true on empty, as `/sys/bus/pci/devices/` will classify as existing.